python/lsst/ts/tcpip/client.py

```diff
@@ -0,0 +1,262 @@
+        )
+        if self._was_connected != connected:
+            self._was_connected = connected
+            if self.__connect_callback is not None:
+                try:
+                    await self.__connect_callback(self)
+                except Exception:
+                    self.log.exception("connect_callback failed.")
+
+    async def basic_close(self) -> None:
+        """Close the connected client socket, if any.
+
+        Like `close` except does not clear ``self.should_be_connected``,
+        nor cancel ``self._monitor_connection_task``.
+        """
+        try:
+            self.log.info("Closing")
+            if self.writer is None:
+                return
+
+            writer = self.writer
+            self.writer = None
+            await utils.close_stream_writer(writer)
+        except Exception:
+            self.log.exception("close failed; continuing")
+        finally:
+            await self.call_connect_callback()
+            if self.done_task.done():
+                self.done_task.set_result(None)
+
+    async def close(self) -> None:
+        """Close the connected client socket, if any.
+
+        Call connect_callback if a client was connected.
+        """
+        self.should_be_connected = False
+        self._monitor_connection_task.cancel()
+        await self.basic_close()
+
+    async def start(self, **kwargs: typing.Any) -> None:
+        """Connect to the TCP/IP server.
+
+        This is called automatically by the constructor,
+        and is not intended to be called by the user.
+        It is a public method so that subclasses can override it.
+
+        Parameters
+        ----------
+        **kwargs : dict[str, Any]
+            Additional keyword arguments for `asyncio.open_connection`,
+            beyond host and port.
+
+        Raises
+        ------
+        RuntimeError
+            If start already called.
+        """
+        if self._started:
+            raise RuntimeError("Start already called")
+
+        self._started = True
+        self.reader, self.writer = await asyncio.open_connection(
+            host=self.host, port=self.port, **kwargs
+        )
+        self.should_be_connected = True
+        if self.monitor_connection_interval > 0:
+            self._monitor_connection_task = asyncio.create_task(
+                self._monitor_connection()
+            )
+        await self.call_connect_callback()
+        self.log.info(f"Client connected to host={self.host}; port={self.port}")
+
+    async def _monitor_connection(self) -> None:
+        """Monitor to detect if the other end drops the connection.
+
+        Raises
+        ------
+        RuntimeError
+            If self.monitor_connection_interval <= 0
+        """
+        if self.monitor_connection_interval <= 0:
+            raise RuntimeError(f"Bug: {self.monitor_connection_interval=} <= 0")
+        while self.connected:
+            await asyncio.sleep(self.monitor_connection_interval)
```

Comment:
In my own test, I felt the check of `self.connected` might not be enough. In `ts_m2com`, I also check the `ConnectionError` and `asyncio.IncompleteReadError`. But I admit the above two will only happen when you try to read something from the pipe:

https://github.com/lsst-ts/ts_m2com/blob/develop/python/lsst/ts/m2com/tcp_client.py#L185-L201

The point is that the ` _monitor_connection()` is very tricky because some errors might only happen when you try to read or write something in pipes.

Reply:
Yes. I agree, and that is a very good point. If this class wrapped the read methods then that wrapper could handle those exceptions and update the connection state. But there are so many ways to read (including read_into), that I have been reluctant to wrap those methods.

For now I would rather that users detect those two exceptions and call `client.basic_close` if they are seen -- as per the example in doc/index.rst.

Reply:
Thinking about this more I have decided to wrap the read and write methods, and thus manage exceptions that indicate the socket has closed. This frees the user from having to tell the OneClientServer or Client that the connection has closed.